src/mpi/coll/iallreduce/iallreduce_tsp_recexch.c

```diff
@@ -9,6 +9,501 @@
 #include "iallreduce_tsp_recursive_exchange_common.h"
 #include "mpir_misc.h"
 
+/* Allreduce single buffer and multiple buffer without dtcopy algorithms are an optimization to the existing
+ * allreduce recexch algorithms. These two algorithms are decoupled and written as separate algorithms
+ * for readability */
+
+/* Routine to schedule a recursive exchange based allreduce using single receive buffer without
+ * data copy in step2. Similar to other recursive exchange algorithms, this algorithm also involves "k"
+ * partners in every communication phase. Removing the data copy in step2 improves performance but at the cost
+ * of increased dependencies. */
```

Comment:
What are the increased dependencies? Dependencies in the schedule graph? If it always improves performance, then I think it should *replace* the old algorithm.

Reply:
recursive exchange algorithm may have multiple phases of exchanging messages with neighbors. If to avoid temporary buffers, we will need to make sure the previous phase finished before we can reuse the buffer, therefore this creates more dependence in the TSP scheduler. Because of this it is a trade-off compared to the data copy scheme which allocates temporary buffer however potentially allows more overlapping between phases.